parmys-plugin/parmys.cc

```diff
@@ -0,0 +1,1175 @@
+/*
+ * Copyright 2022 Daniel Khadivi
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include "kernel/celltypes.h"
+#include "kernel/yosys.h"
+
+#include <regex>
+
+#include "netlist_utils.h"
+#include "odin_globals.h"
+#include "odin_ii.h"
+#include "odin_util.h"
+
+#include "vtr_memory.h"
+#include "vtr_path.h"
+#include "vtr_util.h"
+
+#include "netlist_check.h"
+
+#include "partial_map.h"
+
+#include "netlist_visualizer.h"
+
+#include "parmys_resolve.hpp"
+
+#include "BlockMemories.hpp"
+#include "adders.h"
+#include "arch_util.h"
+#include "hard_blocks.h"
+#include "memories.h"
+#include "multipliers.h"
+#include "netlist_cleanup.h"
+#include "netlist_statistic.h"
+#include "netlist_visualizer.h"
+#include "read_xml_config_file.h"
+#include "subtractions.h"
+
+#include "ast_util.h"
+#include "parmys_update.hpp"
+#include "parmys_utils.hpp"
+
+USING_YOSYS_NAMESPACE
+PRIVATE_NAMESPACE_BEGIN
+
+#define GND_NAME "$false"
+#define VCC_NAME "$true"
+#define HBPAD_NAME "$undef"
+
+struct Bbox {
+    std::string name;
+    std::vector<std::string> inputs, outputs;
+};
+
+CellTypes ct;
+
+struct ParMYSPass : public Pass {
+
+    static void hook_up_nets(netlist_t *odin_netlist, Hashtable *output_nets_hash)
+    {
+        nnode_t **node_sets[] = {odin_netlist->internal_nodes, odin_netlist->ff_nodes, odin_netlist->top_output_nodes};
+        int counts[] = {odin_netlist->num_internal_nodes, odin_netlist->num_ff_nodes, odin_netlist->num_top_output_nodes};
+        int num_sets = 3;
+
+        int i;
```

Comment:
why not declare this in the loop definition?

Reply:
refined the `for` loop syntax within the whole code base of the plugin.